src/pairing.ts

```diff
@@ -0,0 +1,305 @@
+    this.receiver.subscribe(decoder);
+
+    // the received reads the initiator's payloads, and returns the (decrypted) transport message the initiator sent
+    let hsMessage = await this.receiver.nextMessage(this.contentTopic);
+
+    // Note that the received verifies if the received payloadV2 has the expected messageNametag set
+    let step = this.handshake.stepHandshake({
+      readPayloadV2: hsMessage.payloadV2,
+      messageNametag: this.qrMessageNameTag,
+    });
+
+    const initiatorCommittedStaticKey = new Uint8Array(step.transportMessage);
+
+    const hsContentTopic = WakuPairing.toCustomContentTopic(this.qr, "TODO-HSContentTopic"); // TODO: how to calculate ct-id?
+
+    // Subscribe to new contentTopic
+    decoder = new NoiseHandshakeDecoder(hsContentTopic);
+    this.receiver.subscribe(decoder);
+
+    this.eventEmitter.emit("authCodeGenerated", this.handshake.genAuthcode());
+
+    const confirmed = await this.isAuthCodeConfirmed();
+    if (!confirmed) {
+      throw new Error("authcode is not confirmed");
+    }
+
+    // 2nd step
+    // <- sB, eAsB    {r}
+    // Receiver writes and returns a payload
+    step = this.handshake.stepHandshake({
+      transportMessage: this.randomFixLenVal,
+      messageNametag: this.handshake.hs.toMessageNametag(),
+    });
+
+    // We prepare a Waku message from receiver's payload2
+    const encoder = new NoiseHandshakeEncoder(hsContentTopic, step);
+    await this.sender.publish(encoder, {});
+
+    // 3rd step
+    // -> sA, sAeB, sAsB  {s}
+
+    hsMessage = await this.receiver.nextMessage(hsContentTopic);
+
+    // The receiver reads the initiator's payload sent by the initiator
+    const hsStep = this.handshake.stepHandshake({
+      readPayloadV2: hsMessage.payloadV2,
+      messageNametag: this.handshake.hs.toMessageNametag(),
+    });
+
+    if (!this.handshake.hs.rs) throw new Error("invalid handshake state");
+
+    // The receiver further checks if the initiator's commitment opens to the initiator's static key received
+    const expectedInitiatorCommittedStaticKey = commitPublicKey(this.handshake.hs.rs, hsStep.transportMessage);
+    if (!uint8ArrayEquals(expectedInitiatorCommittedStaticKey, initiatorCommittedStaticKey)) {
+      throw new Error("expected committed static key does not match the initiator actual committed static key");
+    }
+
+    // Secure Transfer Phase
+    const hsResult = this.handshake.finalizeHandshake();
+
+    return this.getSecureCodec(hsResult);
+  }
+
+  private getSecureCodec(hsResult: HandshakeResult): [NoiseSecureTransferEncoder, NoiseSecureTransferDecoder] {
+    const secureContentTopic = WakuPairing.toCustomContentTopic(this.qr, "TODO-Secure-ContentTopic"); // TODO: how to calculate ct-id?
+    const secureEncoder = new NoiseSecureTransferEncoder(secureContentTopic, hsResult);
+    const secureDecoder = new NoiseSecureTransferDecoder(secureContentTopic, hsResult);
+
+    return [secureEncoder, secureDecoder];
+  }
```

Comment:
@s1fr0, @D4nte  In this step I create the codecs necessary to encode/decode messages using the cipherstates obtained from calling `finalizeHandshake`. The way js-waku works now is to have a content topic associated to a encoder/decoder, however I read in https://rfc.vac.dev/spec/37/ the following:

> In order to reduce users’ metadata leakage, it is desirable (as suggested in [35/WAKU2-NOISE](https://rfc.vac.dev/spec/35/#after-handshake)) that content topics used for communications change every time a new message is exchanged. This can be easily realized by employing a key derivation function to compute a new session-id from the previously employed one (e.g. session-id = HKDF(prev-session-id)), while keeping the Inbound/outbound Cipher States, the content topic derivation mechanism and the stale mechanism the same as above. In this case, when one party sends and receives at least one message, he SHALL publicly mark as stale all Noise sessions relative to messages exchanged before the earlier of these two send/receive events.

Is this meant to be used after `finalizeHandshake` is executed? if yes, @fryorcraken, is this logic an application concern? or should be something js-waku should provide? I imagine we should have some sort of state machine shared between the codecs to know when to switch content topics as well as notifying filter about any new content topic being used. 

The rfc also uses the term `it is desirable`, does this mean that initially we could use a single content topic and in future versions do the content topic change every time a new message is exchanged?

Reply:
Sorry took me some times to go through notifications.

It seems that we want to move away from the frequent content topic change which means the Encoder/Decoder design still fits. Let me know if I am missing something.